docs/cudf/source/developer_guide/testing.md

```diff
@@ -0,0 +1,150 @@
+This _does not_ include e.g. cuDF or pandas objects.
+
+With that in mind, here are some ground rules for how to parametrize.
+
+Use `pytest.mark.parametrize` when:
+- One test must be run on many inputs and those inputs are simple to construct.
+
+Use fixtures when:
+- One or more tests must be run on the same set of inputs,
+  and all of those inputs can be constructed with simple parametrizations.
+  In practice, that means that it is acceptable to use a fixture like this:
+  ```python
+      @pytest.fixture(params=["a", "b"])
+      def foo(request):
+          if request.param == "a":
+              # Some complex initialization
+          elif request.param == "b":
+              # Some other complex initialization
+  ```
+  In other words, the construction of the fixture may be complex,
+  as long as the parametrization of that construction is simple.
+- One or more tests must be run on the same set of inputs,
+  and at least one of those inputs requires complex parametrizations.
+  In this case, the parametrization of a fixture should be decomposed
+  by using fixtures that depend on other fixtures.
+  ```python
+      @pytest.fixture(params=["a", "b"])
+      def foo(request):
+          if request.param == "a":
+              # Some complex initialization
+          elif request.param == "b":
+              # Some other complex initialization
+
+      @pytest.fixture
+      def bar(foo):
+         # do something with foo like initialize a cudf object.
+
+      def test_some_property(bar):
+          # will be run for each value of bar that results from each value of foo.
+          assert some_property_of(bar)
+  ```
+
+#### Complex parametrizations
+
+The lists above document common use cases.
+However, more complex cases may arise.
+One of the most common alternatives is where, given a set of test cases,
+different tests need to run on different subsets with a nonempty intersection.
+Fixtures and parametrization are only capable of handling the Cartesian product of parameters,
+i.e. "run this test for all values of `a` and all values of `b`".
+
+There are multiple potential solutions to this problem.
+One possibility is to encapsulate common test logic in a helper function,
+then call it from multiple `test_*` functions that construct the necessary inputs.
+Another possibility is to use functions rather than fixtures to construct inputs, allowing for more flexible input construction:
+```python
+def get_values(predicate):
+    values = range(10)
+    yield from filter(predicate, values)
+
+def test_evens():
+    for v in get_values(lambda x: x % 2 == 0):
+        # Execute test
+
+def test_odds():
+    for v in get_values(lambda x: x % 2 == 1):
+        # Execute test
+```
+
+Other approaches are also possible, and the best solution should be discussed on a case-by-case basis during PR review.
+Developers should avoid performing GPU memory allocations during test collection.
```

Comment:
This sentence is, I think, important enough that we shouldn't bury it at the very end of this section. Perhaps move it after the sentence that says "This does not include CUDF or pandas objects" ?

Reply:
Good call.